internal/security/config/command/proxy/tls/command.go

```diff
@@ -61,9 +83,235 @@ func NewCommand(
+		return nil, fmt.Errorf("Failed to read private key from file %s: %w", c.privateKeyPath, err)
+	}
+	return &bootstrapConfig.CertKeyPair{Cert: string(certPem), Key: string(prvKey)}, nil
+}
+
+func (c *cmd) uploadProxyTlsCert() error {
+	// try to read both files and make sure they are existing
+	certKeyPair, err := c.readCertKeyPairFromFiles()
+	if err != nil {
+		return err
+	}
+
+	// to see if any proxy certificates already exists
+	// if yes, then delete them all first
+	// and then upload the new TLS certificate
+	existingCertIDs, err := c.listKongTLSCertificates()
+	if err != nil {
+		return err
+	}
+
+	c.loggingClient.Debug(fmt.Sprintf("number of existing Kong tls certs = %d", len(existingCertIDs)))
+
+	if len(existingCertIDs) > 0 {
+		// delete the existing certs first
+		// ideally, it should only have one certificate to be deleted
+		// Disclaimer: Kong TLS certificate should only be uploaded via secret-config utility
+		for _, certID := range existingCertIDs {
+			if err := c.deleteKongTLSCertificateById(certID); err != nil {
+				return err
+			}
+		}
+	}
+
+	// post the certKeyPair as the new one
+	if err := c.postKongTLSCertificate(certKeyPair); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *cmd) listKongTLSCertificates() ([]string, error) {
+	// list snis certificates association to see if any already exists
+	certKongURL := strings.Join([]string{c.configuration.KongURL.GetProxyBaseURL(), "snis"}, "/")
+	c.loggingClient.Info(fmt.Sprintf("list snis tls certificates on the endpoint of %s", certKongURL))
+	req, err := http.NewRequest(http.MethodGet, certKongURL, http.NoBody)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to prepare request to list Kong snis tls certs: %w", err)
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to send request to list Kong snis tls certs: %w", err)
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read response body to list Kong snis tls certs: %w", err)
+	}
+
+	var parsedResponse map[string]interface{}
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		if err := json.NewDecoder(bytes.NewReader(responseBody)).Decode(&parsedResponse); err != nil {
+			return nil, fmt.Errorf("Unable to parse response from list snis certificate: %w", err)
+		}
+	default:
+		return nil, fmt.Errorf("List Kong tls snis certificates request failed with code: %d", resp.StatusCode)
+	}
+
+	var jsonData []byte
+	jsonData, err = json.Marshal(parsedResponse["data"])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to json marshal parsed response data: %w", err)
+	}
+
+	// the list snis get API returns an array of snis-certs association
+	var parsedSnisCertData []map[string]interface{}
+	if err := json.NewDecoder(bytes.NewReader(jsonData)).Decode(&parsedSnisCertData); err != nil {
+		return nil, fmt.Errorf("Unable to parse response for parsed SNIS cert data: %w", err)
+	}
```

Comment:
Should only need one Decode, and not marshal and decode again if struct is defined appropriately.

What does the raw JSON response look like?

Reply:
not a fan to this rule. see the structure in test for the raw JSON response.

Reply:
Hi @lenny-intel: 

example of the first time response: 
```json
 {"next":null,"data":[]}
```

after post some cert successfully, example of the second time response: 
```json
{"next":null,"data":[{"created_at":1607534629,"id":"646c762c-f86f-4222-a7bb-acb27d108cd2","tags":null,"name":"localhost","certificate":{"id":"6e8dbc72-d175-4e47-8064-9a45634a822b"}},{"created_at":1607534629,"id":"70f4aa8c-94c7-4985-881d-ab915e30c663","tags":null,"name":"kong","certificate":{"id":"6e8dbc72-d175-4e47-8064-9a45634a822b"}}]}
```

there is "next" on the first level and thus `map[string]interface{}`
but there is "certificate" object inside `data` after the first level, so the `certificate`'s id is nested inside the `data`:`certificate` and thus requires the second time of decoding.

We can change the mechanism to do marshal/unmarshal the predefined structures but there is two drawbacks to do this:
 1. The json user-defined structure is "Globally" exposed (for json to marshal/unmarshal to work) which is undesired in this utility.
 1. Inconsistent with other implementation of subcommands: the other existing subcommnads like adduser/deluser, what have you, currently don't implement this way; so using coding/encoding is to make consistent with the implementation of other subcommands.  For the sake of consistence, then will have to change all codes, and this is beyond the scope of this PR.